File: brain/nowplaying/applemusic.py
"""Apple Music adapter — mirrors the proven now-playing.jsx widget stack.

Apple has no server-side currently-playing API, so this layers three sources
(first with an answer wins), exactly like the Übersicht widget:

  0. now-spinning state.json (Group Container) — companion-app push state,
     used when present and playing.
  1. osascript against Music.app — real-time local playback (incl. AirPlay
     FROM the Mac), playing or paused. Never launches Music.
  2. musickit-fetch.py --nowplaying — the account's most-recently-played
     track across ALL devices (this is what catches the iPhone), using the
     MusicKit credentials already set up in ~/.config/widgetsuite/.

Artwork is URL-based (MusicKit catalog --cover, then the iTunes Search
ladder) — never AppleScript raw bytes — so art/fetch.py caches it cleanly.

Modes:
  endpoint = ""                          local: brain runs on this Mac
  endpoint = "http://<mac>.local:8787"   remote: brain on the Pi polls
                                         scripts/mac_reporter.py on the Mac
"""
import json
import os
import subprocess
import sys
import time

from . import NowPlaying, NowPlayingSource

WIDGETSUITE = os.path.expanduser("~/.config/widgetsuite")
MUSICKIT_HELPER = os.path.join(WIDGETSUITE, "musickit-fetch.py")
STATE_JSON = os.path.expanduser(
    "~/Library/Group Containers/group.com.jalenedusei.widgetsuite/"
    "now-spinning/state.json")
# A crashed companion leaves state.json claiming "playing" forever, which
# would pin the wall on a dead track and shadow the live tiers below it
# (the reporter's push tier expires for exactly this reason). The file is
# rewritten on track changes, so 20 minutes of silence outlasts almost any
# song while still letting a corpse expire.
STATE_JSON_TTL_S = 20 * 60

# NB: "st" is a reserved word in macOS 27 AppleScript — hence pstate.
OSA_META = '''
if application "Music" is not running then return "state=not_running"
tell application "Music"
  set pstate to (player state as text)
  if pstate is not in {"playing", "paused"} then return "state=" & pstate
  set t to current track
  set pos to "?"
  set dur to "?"
  try
    set pos to (player position as text)
  end try
  try
    set dur to (duration of t as text)
  end try
  return "state=" & pstate & linefeed & (persistent ID of t) & linefeed & (name of t) & linefeed & (artist of t) & linefeed & (album of t) & linefeed & pos & linefeed & dur
end tell
'''


def _helper_python():
    return "/usr/bin/python3" if os.path.exists("/usr/bin/python3") else sys.executable


def _run_helper(*args, timeout=20):
    if not os.path.exists(MUSICKIT_HELPER):
        return ""
    try:
        proc = subprocess.run([_helper_python(), MUSICKIT_HELPER, *args],
                              capture_output=True, text=True, timeout=timeout)
        return proc.stdout.strip()
    except Exception:
        return ""


def _ms(val):
    try:
        return int(float(val) * 1000)
    except (TypeError, ValueError):
        return None


def _osa_now():
    """Local Music.app state -> dict, or None when idle/stopped/not running."""
    try:
        proc = subprocess.run(["osascript", "-"], input=OSA_META,
                              capture_output=True, text=True, timeout=10)
    except subprocess.TimeoutExpired:
        return None
    if proc.returncode != 0:
        err = proc.stderr.strip()
        if "-1743" in err:
            raise RuntimeError(
                "Automation permission denied — System Settings > Privacy & "
                "Security > Automation: allow this app to control Music")
        raise RuntimeError(f"osascript: {err[:160]}")
    out = proc.stdout.rstrip("\n")
    if "\n" not in out:
        return None
    fields = (out.split("\n") + ["?"] * 7)[:7]
    state, pid, title, artist, album, pos, dur = fields
    return {
        "state": state.split("=", 1)[1],
        "pid": pid, "title": title, "artist": artist, "album": album,
        "progress_ms": _ms(pos), "duration_ms": _ms(dur),
    }


def _itunes_art(term, entity):
    try:
        import requests  # lazy: keeps module import stdlib-only for the reporter
        resp = requests.get("https://itunes.apple.com/search",
                            params={"term": term, "entity": entity, "limit": 1},
                            timeout=10)
        resp.raise_for_status()
        results = resp.json().get("results") or []
        url = results[0].get("artworkUrl100") if results else None
        return url.replace("100x100bb", "600x600bb") if url else None
    except Exception:
        return None


class AppleMusicSource(NowPlayingSource):
    name = "applemusic"

    phone_age = None                 # seconds since the phone's last push

    def __init__(self, endpoint: str = ""):
        self.endpoint = (endpoint or "").rstrip("/")
        self._art_key = None
        self._art_url_cached = None
        self._warned = False

    # ---- artwork (the widget's fallback ladder, simplified) ------------
    def _art_url(self, title, artist, album):
        key = (artist, album, title)
        if key == self._art_key:
            return self._art_url_cached
        url = _run_helper("--cover", title, artist, album) or None
        if not url:
            clean_album = album.removesuffix(" - Single").removesuffix(" - EP")
            url = (_itunes_art(f"{artist} {clean_album}", "album")
                   or _itunes_art(f"{artist} {title}", "song")
                   or _itunes_art(title, "song"))
        if url:  # cache hits only, so misses retry next poll
            self._art_key, self._art_url_cached = key, url
        return url

    # ---- tiers ---------------------------------------------------------
    def _tier0_state_json(self):
        try:
            if time.time() - os.path.getmtime(STATE_JSON) > STATE_JSON_TTL_S:
                return None
            with open(STATE_JSON) as fh:
                s = json.load(fh)
        except (FileNotFoundError, json.JSONDecodeError, OSError):
            return None
        if not s.get("playing") or not s.get("track"):
            return None
        return NowPlaying(
            track_id="applemusic:" + (s.get("id")
                                      or f"{s.get('artist')}|{s.get('track')}"),
            title=s["track"], artist=s.get("artist", "?"),
            album=s.get("album", "?"), art_url=s.get("art"),
            progress_ms=None, duration_ms=None, is_playing=True,
        )

    def _tier1_mac(self):
        info = _osa_now()
        if info is None:
            return None, ("", "")
        art = self._art_url(info["title"], info["artist"], info["album"])
        now = NowPlaying(
            track_id=f"applemusic:{info['pid']}",
            title=info["title"], artist=info["artist"], album=info["album"],
            art_url=art, progress_ms=info["progress_ms"],
            duration_ms=info["duration_ms"],
            is_playing=info["state"] == "playing",
        )
        return now, (info["title"], info["artist"])

    def _tier2_account(self, mac_track="", mac_artist=""):
        out = _run_helper("--nowplaying", mac_track, mac_artist)
        if not out:
            return None
        try:
            s = json.loads(out)
        except json.JSONDecodeError:
            return None
        return NowPlaying(
            track_id=f"applemusic:{s.get('artist')}|{s.get('track')}",
            title=s.get("track", "?"), artist=s.get("artist", "?"),
            album=s.get("album", "?"), art_url=s.get("art"),
            progress_ms=None, duration_ms=None, is_playing=True,
        )

    def _local(self):
        now = self._tier0_state_json()
        if now:
            return now
        now, (mac_track, mac_artist) = self._tier1_mac()
        if now and now.is_playing:
            return now
        # Mac idle or paused: the account view catches iPhone playback; the
        # helper stays silent when the account track is just the Mac's paused
        # one, so the paused sleeve keeps the wall.
        return self._tier2_account(mac_track, mac_artist) or now

    def _remote(self):
        import requests  # lazy; the Pi venv has it
        try:
            resp = requests.get(self.endpoint + "/nowplaying", timeout=8)
        except requests.RequestException as exc:
            self.phone_age = None
            if not self._warned:
                print(f"[applemusic] reporter unreachable at {self.endpoint} "
                      f"({exc.__class__.__name__}) — start "
                      "scripts/mac_reporter.py on the Mac")
                self._warned = True
            return None
        self._warned = False
        # How long since the owner's phone last spoke to the reporter.
        # Presence, not music: the away behaviour reads this.
        age = resp.headers.get("X-Phone-Age")
        self.phone_age = float(age) if age is not None else None
        if resp.status_code != 200:
            return None
        try:
            data = resp.json()
        except ValueError:
            return None
        # The reporter decorates its payload (e.g. "tier"); take only the
        # dataclass's own fields so decoration never breaks this side.
        return NowPlaying(**{k: v for k, v in data.items()
                             if k in NowPlaying.__dataclass_fields__})

    def get_current(self):
        return self._remote() if self.endpoint else self._local()

File: web/src/lib/sources/applemusic.ts
import type { NowPlaying, NowPlayingSource, SourceConfig } from "../types";
import { fetchJson, status, TransportError } from "./base";

export const DEFAULT_BRIDGE_URL = "http://localhost:8787/nowplaying";

/** Which side of the Mac reporter answered - the values it actually sends
 * (scripts/mac_reporter.py): a live phone push, the Mac tier ladder, or
 * "Mac: <app>" when another app on the Mac (Spotify, a browser tab on
 * YouTube Music...) was read through its Now Playing. */
export const TIERS: Record<string, string> = {
  "iphone push": "iPhone push (live from the phone)",
  "mac tiers": "Mac tiers (state file, Music.app, MusicKit)",
};

function tierLabel(v: unknown): string {
  if (v == null) return "unreported tier";
  const s = String(v);
  if (s.toLowerCase().startsWith("mac: ")) return `${s} (read from the Mac's Now Playing)`;
  return TIERS[s.toLowerCase()] ?? s;
}

export const appleMusicSource: NowPlayingSource = {
  id: "applemusic",
  label: "Apple Music",
  description:
    "Three tiers on the Mac: companion state file, Music.app over AppleScript, then a MusicKit account query that catches playback on an iPhone.",

  probe(cfg: SourceConfig) {
    const mode = String(cfg.config.mode ?? "bridge");
    if (mode === "direct") {
      return status(
        "applemusic",
        "needs-bridge",
        "Direct mode talks to macOS APIs (AppleScript, MusicKit) that a browser cannot reach. It needs the desktop build; it is unavailable in this browser.",
      );
    }
    const url = String(cfg.config.bridgeUrl ?? "").trim();
    if (!url) {
      return status("applemusic", "needs-setup", "Set the bridge URL of the reporter running on your Mac.");
    }
    return status("applemusic", "disconnected", `Polling ${url}. Not yet contacted.`);
  },

  async getCurrent(cfg: SourceConfig): Promise<NowPlaying | null> {
    const mode = String(cfg.config.mode ?? "bridge");
    if (mode === "direct") {
      throw new TransportError(
        "Direct mode requires the desktop build. This browser cannot call AppleScript or MusicKit.",
        "network",
      );
    }
    const url = String(cfg.config.bridgeUrl ?? DEFAULT_BRIDGE_URL);
    const { status: code, body } = await fetchJson(url);
    if (code === 204) return null;
    if (code >= 400) throw new TransportError(`Bridge returned HTTP ${code}.`, "http", code);
    if (!body || typeof body !== "object") {
      throw new TransportError("Bridge returned a body that is not a now-playing object.", "parse");
    }
    const b = body as Record<string, unknown>;
    // The Mac reporter speaks snake_case (its dataclass), older bridges camelCase.
    const pick = (a: string, c: string) => b[a] ?? b[c];
    const id = pick("trackId", "track_id");
    if (!b.title && !id) {
      throw new TransportError("Bridge response is malformed: it has no title or trackId.", "parse");
    }
    const rawId = String(id ?? `${b.artist ?? ""}-${b.title ?? ""}`);
    const art = pick("artUrl", "art_url");
    const progress = pick("progressMs", "progress_ms");
    const duration = pick("durationMs", "duration_ms");
    const playing = pick("isPlaying", "is_playing");
    return {
      trackId: rawId.startsWith("applemusic:") ? rawId : `applemusic:${rawId}`,
      title: String(b.title ?? "Unknown title"),
      artist: String(b.artist ?? "Unknown artist"),
      album: String(b.album ?? ""),
      artUrl: art ? String(art) : null,
      progressMs: typeof progress === "number" ? progress : null,
      durationMs: typeof duration === "number" ? duration : null,
      isPlaying: playing !== false,
      tier: tierLabel(b.tier),
    };
  },
};
